server/udns_server.ml

```diff
@@ -478,64 +501,69 @@ module Notification = struct
   *)
   let retransmit = Array.map Duration.of_sec [| 1 ; 3 ; 7 ; 20 ; 40 ; 60 ; 180 |]
 
-  let retransmit ns now =
+  let retransmit server ns now ts =
     let max = pred (Array.length retransmit) in
     IPM.fold (fun ip map (new_ns, out) ->
         let new_map, out =
           Domain_name.Map.fold
-            (fun name (ts, count, packet, key) (new_map, out) ->
-               if Int64.add ts retransmit.(count) < now then
+            (fun name (oldts, count, mac, packet, key) (new_map, outs) ->
+               if Int64.sub ts retransmit.(count) > oldts then
+                 let out, mac = encode_and_sign key server now packet in
                  (if count = max then begin
                      Log.warn (fun m -> m "retransmitting notify to %a the last time %a"
                                  Ipaddr.V4.pp ip Packet.pp packet) ;
                     new_map
                    end else
-                    (Domain_name.Map.add name (ts, succ count, packet, key) new_map)),
-                 (ip, fst (Packet.encode `Udp packet)) :: out
+                    (Domain_name.Map.add name (oldts, succ count, mac, packet, key) new_map)),
+                 (ip, out) :: outs
                else
-                 (Domain_name.Map.add name (ts, count, packet, key) new_map, out))
+                 (Domain_name.Map.add name (oldts, count, mac, packet, key) new_map, outs))
             map (Domain_name.Map.empty, out)
         in
         (if Domain_name.Map.is_empty new_map then new_ns else IPM.add ip new_map new_ns),
         out)
       ns (IPM.empty, [])
 
-  let notify conn ns server now zone soa =
-    let remotes = to_notify conn ~data:server.data ~auth:server.auth zone in
-    Log.debug (fun m -> m "notifying %a %a" Domain_name.pp zone
-                  Fmt.(list ~sep:(unit ",@ ")
-                         (pair ~sep:(unit ", key") Ipaddr.V4.pp
-                            (option ~none:(unit "no key") Domain_name.pp)))
-                  (IPM.bindings remotes));
+  let notify_one ns server now ts zone soa ip key =
     let packet =
       let question = Packet.Question.create zone Soa
       and header =
-        (* TODO: all are getting the same ID *)
         let id = Randomconv.int ~bound:(1 lsl 16 - 1) server.rng in
```

Comment:
How about `Randomconv.int16 server.rng`, or does this need to be `0 .. 65534` specifically?
Same here https://github.com/roburio/udns/blob/445035a96f3aab000aa97d84a548e966fe81756e/server/udns_server.ml#L909

65534 (`1 lsl 16 = 0x10000`, `1 lsl 16 - 1 = 0xffff`, bound being exclusive in `Randomconv.int ~bound` meaning you get `0 ... 0xfffe` inclusively) seems like a weird number to me; usually you'd want `0x0 ... 0xffff`?